Solutions/Menes.Abstractions/Menes/Auditing/AuditLog.cs

```diff
@@ -30,12 +39,12 @@ public class AuditLog
         /// <summary>
         /// Gets or sets the Id of the tenant for the operation.
         /// </summary>
-        public string TenantId { get; set; }
+        public string? TenantId { get; set; }
```

Comment:
Should these actually be not-nullable?

Reply:
Well... the two places that set this both retreive the `TenantId` from an `IOpenApiContext`. So this is really a question about `IOpenApiContext.TenantId`.

There are two reasons this is currently nullable:

- the code in Menes that sets this doesn't run until after the parameter builder has run (and just before the access policies are checked), so it's definitely true to say that there are phases in the processing of a request in which `IOpenApiContext` cannot possibly be non-null
- the code that sets this doesn't always run

If only the first of these were an issue, we could argue that because audit logs are only generated after execution, the tenant ID definitely should have been set by the time we get as far as creating the `AuditLog`, and so this should be non-nullable as you say.

However, it's that second point that's the problem. We have this in `OpenApiOperationInvoker.cs`:

```csharp
// Try to find the tenantID in the parameters, but before we check the access policies.
if (namedParameters.TryGetValue("tenantId", out object tenantIdObject) && tenantIdObject is string tenantId)
{
    context.CurrentTenantId = tenantId;
}
```

So we have basically baked into the operation invoker the idea that the tenant ID can be found by looking for an argument named `tenantId`.

Is that always present in Marain services? So far, yes.

Will this always be present in anything that uses `Menes`? Good question. Currently, the answer is that no, it doesn't have to be. (The PetStore demo doesn't have this, for example.) By changing this property to a non-nullable string, we'd be imposing the requirement that there would always have to be a named parameter called `tenantId`.

So that would be a change in behaviour. The question, then, is: do we want to add that constraint?

Reply:
And no, it shouldn't be. I really want to remove the baked-in notion of tenancy and have it more layered up; the fact that the auditlog has a baked in notion of tenancy is "not a happy thing" in the first place.